Introduction to Hooks (#1545)
enkidevs/curriculum: javascript/react/introduction-to-hooks/what-are-hooks.md

---
author: kapnobatai136

levels:

  - medium
  
aspects:

  - new

  - workout

  - deep

type: normal

category: must-know

links:
  - '[hooks-intro](https://reactjs.org/docs/hooks-intro.html){documentation}'

---

# What are hooks?

---
## Content

> Hooks represent a new way of using state with your functional components. They are the best approach at reusing stateful logic and enable you to use more of React's features without having to use classes.

Reviewer: Saying they are best sounds a bit subjective. How about this?

```suggestion
Hooks represent a new way of using state with your functional components. They are the great for reusing stateful logic and enable you to use more of React's features without having to use classes.
```

There are three main types of hooks:
- State Hooks
- Effect Hooks
- Custom Hooks

> Before hooks, your components would start out simple but with time they would grow into a random grouping of stateful logic and side effects. Your components could perform some data fetching in the `componentDidMount` or `componentDidUpdate` lifecycle methods, but the `componentDidMount` could also contain some event listeners that were totally unrelated. Hooks introduced a way of splitting up components into smaller functions such that each part is related.

Reviewer: This is a good point.

Here's a quick revision:

```suggestion
Before hooks, many initially simple React component would eventually grow to contain many unrelated groupings of stateful logic and side effects. A component could perform some data fetching in the `componentDidMount` or `componentDidUpdate` lifecycle methods, while the `componentDidMount` could also contain some event listeners. Hooks allow us to separate these unrelated parts by encapsulating related logic into functions.
```

Now, let's take a look at a simple state hook:

```jsx
import React, { useState } from "react";

function Enki() {
  const [count, setCount] = useState(0);

  return (
    <div>
      <p>You clicked {count} times</p>
      <button
        onClick={() => setCount(count + 1)}
      >
        Click me
      </button>
    </div>
  );
}
```

What the `Enki` component does is that every time the button is clicked, the `count` state is incremented by one. Now, if we would've used class components, the code would've looked something like this:

```jsx

> class Enki extends React.Component {

Reviewer: We might need a better example here (and above) to better support the [point about grouping unrelated logic](https://github.com/enkidevs/curriculum/pull/1545/files#r326823090)

  constructor(props) {
    super(props);
    this.state = {
      count: 0
    };
  }

  render() {
    return (
      <div>
        <p>
          You clicked {this.state.count}
          {" "}times
        </p>
        <button
          onClick={() =>
            this.setState({

>               count: this.state.count + 1

Reviewer 2: You might want to have
```
onClick={() => this.setState(state => ({count: state.count + 1}))}
```
as per https://reactjs.org/docs/state-and-lifecycle.html#state-updates-may-be-asynchronous

            }))
          }
        >
          Click me
        </button>
      </div>
    );
  }
}
```

> Both components have the same functionality, but you can already see how much easier it is to write state hooks compared to using state in a class component.

Reviewer: ...and then here we can summarize how with hooks related stateful logic and side effects are grouped together and separated from unrelated code.
